src/coreEnforcer.ts

```diff
@@ -279,12 +279,17 @@ export class CoreEnforcer {
       throw new Error('Unable to find policy_effect in model');
     }
 
-    const matcherKey = `${asyncCompile ? 'ASYNC[' : 'SYNC['}${expString}]`;
+    const HasEval: boolean = hasEval(expString);
+    let expression: any;
 
-    let expression = this.matcherMap.get(matcherKey);
-    if (!expression) {
-      expression = asyncCompile ? compileAsync(expString) : compile(expString);
-      this.matcherMap.set(matcherKey, expression);
+    if (!HasEval) {
```

Comment:
Extract matcher `type Matcher = ((context: object) => Promise<any>) | ((context: object) => any);`

Change `private matcherMap: Map<string, ((context: object) => Promise<any>) | ((context: object) => any)> = new Map();` to `private matcherMap: Map<string, Matcher> = new Map();`

then extract code of line 285-295 to `private  getExpression(exp: string): Matcher` for avoid same code in project.

Try to avoid `any` :(

Reply:
>type Matcher = ((context: object) => Promise<any>) | ((context: object) => any);

I am facing a problem doing this.

Reply:
>Try to avoid any :(

I agree with you but sometimes ```any``` is better option, as it sometimes decreases unwanted verbosity of the code. Many times by avoiding ```any``` from the code, we introduce many unwanted control structures.

And also there are many instances where ```any``` is used in the code.

Reply:
> > type Matcher = ((context: object) => Promise) | ((context: object) => any);
> 
> I am facing a problem doing this.

What problem?

Reply:
Alright, solved.